py-development.Rmd

```diff
@@ -0,0 +1,540 @@
+```
+
+> Note: When we defined the function `sum_two_numbers()`, we referred to `num1`
+> and `num2` as *parameters*, while we refer to the numbers we pass to the
+> function call (`2` and `5` above) as *arguments*. Although this might sound
+> confusing at first, it is a standard followed by many programming languages
+> so it is useful to get accustomed to this terminology.
+
+#### Function composition {#py-dev-function-composition}
+
+Combining functions is referred to as function composition. This practice
+allows us to write functions that perform one specific task and then combine
+them for more complicated tasks, which makes code more readable and easier to
+debug. By composing a function from the built-in `len()` and `sum()` functions,
+we can create a more succinct and easier to read version of our previous
+function `sum_and_len_all_numbers()`.
+
+```{python}
+def sum_and_len(list_of_numbers):
+    return sum(list_of_numbers), len(list_of_numbers)
+```
+
+#### Defining default values {#py-dev-defining-default-values}
+
+Up until now, our function calls have included enough arguments to assign one
+to each of the function parameters. This can get tedious in functions with many
+parameters and it is often beneficial to define default values for several of
+the function parameters. As an example, we can modify `sum_two_numbers()` to
+optionally return the two input arguments.
+
+```{python}
+def sum_two_numbers(num1, num2, return_input=False):
+    if return_input:
+        return num1, num2, num1 + num2
+    else:
+        return num1 + num2
+```
+
+By default the function will work just as previously.
+
+```{python}
+sum_two_numbers(2, 5)
+```
+
+But we now also have the option to return the input numbers.
+
+```{python}
+sum_two_numbers(2, 5, return_input=True)
+```
+
+Since the arguments are given in the same order as in the function definition,
+we could have left out the `return_input=` part and just written `True` in the
+third position. If we want to supply the arguments out of order, we need to
+specify the parameter name of each argument.
+
+#### Function documentation {#py-dev-function-documentation}
+
+Functions might appear self-explanatory when they are being written, but it is
+essential that there is proper documentation describing what the function does
+and what types of arguments should be in the input. This helps other people who
+are reading your code and also your future self that will be reusing these
+functions.
+
+In Python, a function is documented in its docstring, which is a multiline
+Python string immediately following the function definition. It is surrounded
+by triple quotes (either single or double) and can look like this.
+
+```{python}
+def sum_two_numbers(num1, num2):
+    '''
+    Add two numbers
+
+    Args:
+        num1 (int, float): The first number to be added.
+        num2 (int, float): The second number to be added.
+
+    Returns:
+        int, float: Sum of the two numbers.
+    '''
```

Comment:
Is there an easy way to add this docstring? E.g. like a key binding (for instance, Ctrl-Shift-R or whatever in RStudio for using roxygen comments).

Reply:
Yes, good point I will add this. There is actually a keybinding, a menu item, and an autocomplete template ([gif here](https://github.com/spyder-ide/spyder/pull/8700))!